checker/src/org/checkerframework/checker/index/OffsetDependentTypesHelper.java

```diff
@@ -0,0 +1,89 @@
+package org.checkerframework.checker.index;
+
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.util.TreePath;
+import org.checkerframework.checker.index.upperbound.OffsetEquation;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.treeannotator.TreeAnnotator;
+import org.checkerframework.framework.util.FlowExpressionParseUtil;
+import org.checkerframework.framework.util.FlowExpressionParseUtil.FlowExpressionContext;
+import org.checkerframework.framework.util.dependenttypes.DependentTypesError;
+import org.checkerframework.framework.util.dependenttypes.DependentTypesHelper;
+import org.checkerframework.framework.util.dependenttypes.DependentTypesTreeAnnotator;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * Dependent type helper for array offset expressions. Each array offset expression may be the
+ * addition or subtraction of several Java expressions. For example, {@code array.length - 1}.
+ */
+public class OffsetDependentTypesHelper extends DependentTypesHelper {
+    public OffsetDependentTypesHelper(AnnotatedTypeFactory factory) {
+        super(factory);
+    }
+
+    @Override
+    protected String standardizeString(
+            final String expression,
+            FlowExpressionContext context,
+            TreePath localScope,
+            boolean useLocalScope) {
+        if (DependentTypesError.isExpressionError(expression)) {
+            return expression;
+        }
+        if (indexOf(expression, '-', '+', 0) == -1) {
+            return super.standardizeString(expression, context, localScope, useLocalScope);
+        }
+
+        OffsetEquation equation = OffsetEquation.createOffsetFromJavaExpression(expression);
+        if (equation.hasError()) {
+            return equation.getError();
+        }
+        try {
+            equation.standardizeAndViewpointAdaptExpressions(context, localScope, useLocalScope);
+        } catch (FlowExpressionParseUtil.FlowExpressionParseException e) {
+            return new DependentTypesError(expression, e).toString();
+        }
+
+        return equation.toString();
```

Comment:
It's surprising that `super.standardizeString` is not called (either before or after the main logic of this method), which is inconsistent with line 35 above.

Reply:
I didn't write this code, but `super.standardizeString` is not called because the work of parsing is done in `equation.standardizeAndViewpointAdaptExpressions` for individual terms. I added comments to explain this.